### PR TITLE
Added saving query

### PR DESCRIPTION
**Pull request by smartguest**
This PR adds in the renameQuery call from this PR: https://github.com/microsoft/sqltoolsservice/pull/1227 and is intended to be merged after that PR.

This PR also updates connection objects and the existing queryrunner object to use the new uri, and prevents dispose of the editor during a save, Saving now returns a FileQueryEditorInput directly. To handle any VS import rules, UntitledQueryEditorInput has been moved to sql/base/query/common.

This is for fixing issue #1876

### CODE REVIEW

File: src/sql/workbench/common/editor/query/queryResultsInput.ts [hunk 1]
```diff
@@ -88,6 +88,10 @@ export class QueryResultsInput extends EditorInput {
 		return this._uri;
 	}
 
+	set uri(newUri: string) {
```
[the remark this answers is not in the document]
Comment (smartguest) [reply src/sql/workbench/common/editor/query/queryResultsInput.ts:91]: I would, but uri is already in editorinput, so I can't name it that way.

File: extensions/liveshare/src/constants.ts [hunk 2]
```diff
@@ -33,5 +33,6 @@ export const runQueryAndReturnRequest = 'runQueryAndReturn';
 export const parseSyntaxRequest = 'parseSyntax';
 export const getQueryRowsRequest = 'getQueryRows';
 export const disposeQueryRequest = 'disposeQuery';
+export const notifyConnectionUriChangedRequest = 'notifyConnectionUriChanged';
```
Comment (Charles-Gagnon) [inline extensions/liveshare/src/constants.ts:36]: `connectionUriChangedNotification`

File: extensions/liveshare/src/providers/queryProvider.ts [hunk 3]
```diff
@@ -132,6 +136,12 @@ export class QueryProvider {
 			}]);
 		};
 
+		let notifyConnectionUriChanged = (ownerUri: string): Thenable<void> => {
+			return self._sharedServiceProxy.request(constants.notifyConnectionUriChangedRequest, [{
```
Comment (Charles-Gagnon) [inline extensions/liveshare/src/providers/queryProvider.ts:140]: This doesn't look correct - you should be using `.notify` instead of request and the name should be the actual name of the notification (which should be `connectionUriChanged`, not `notifyConnectionUriChanged`)

File: src/sql/azdata.proposed.d.ts [hunk 4]
```diff
@@ -940,4 +940,11 @@ declare module 'azdata' {
 		 */
 		parentTypeName?: string;
 	}
+
+	export interface QueryProvider {
+		/**
+		 * Change the URI associated with the specified connection.
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:946]: `Notify clients that the URI for a connection has been changed`

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 5]
```diff
@@ -44,6 +48,26 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
 		return this.text.model.hasAssociatedFilePath;
 	}
 
+	override async save(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		let preProcessed = await this.text.save(group, options);
```
Comment (Charles-Gagnon) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:52]: I'd name this something like `fileEditorInput` instead, "preprocessed" isn't really an accurate name here (there's nothing being "processed")

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 6]
```diff
@@ -44,6 +48,26 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
...
+		this._results.uri = newUri;
+		await this.notifyConnectionUriChanged(newUri);
+		let newInput = this.instantiationService.createInstance(FileQueryEditorInput, '', (preProcessed as FileEditorInput), this.results);
+		newInput.state.setState(this.state);
+		return newInput;
+	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		let preProcessed = await this.text.saveAs(group, options);
+		let newUri = preProcessed.resource.toString(true);
```
Comment (Charles-Gagnon) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:63]: These last lines are duplicated between the two save functions, would be nice to combine them.

File: src/sql/platform/connection/common/connectionStatusManager.ts [hunk 7]
```diff
@@ -49,6 +49,19 @@ export class ConnectionStatusManager {
 		return !!this.findConnection(id);
 	}
 
+	public changeConnectionUri(newId: string, oldId: string) {
+		let info = this.findConnection(oldId);
+		if (!info) {
+			this._logService.error(`No connection found associated with old URI : '${oldId}'`);
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionStatusManager.ts:55]: This should probably throw as well (or at the very least it needs to return to avoid null ref on line 57)

File: src/sql/platform/connection/common/connectionStatusManager.ts [hunk 8]
```diff
@@ -49,6 +49,19 @@ export class ConnectionStatusManager {
...
+		}
+		info.ownerUri = newId;
+		if (this._connections[newId]) {
+			this._logService.error(`New URI : '${newId}' is already in the connections list.`);
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionStatusManager.ts:59]: This check should come before updating the info - since we don't want to update the URI if either case is true (also same as above, this should probably throw)

File: src/sql/workbench/api/common/sqlExtHost.protocol.ts [hunk 9]
```diff
@@ -193,6 +193,11 @@ export abstract class ExtHostDataProtocolShape {
 	 */
 	$disposeQuery(handle: number, ownerUri: string): Thenable<void> { throw ni(); }
 
+	/**
+	 * Changes the uri associated with a query
```
Comment (Charles-Gagnon) [inline src/sql/workbench/api/common/sqlExtHost.protocol.ts:197]: `Notifies the client that the URI associated with a connection has changed`

File: src/sql/workbench/common/editor/query/queryEditorInput.ts [hunk 10]
```diff
@@ -106,6 +106,14 @@ export class QueryEditorState extends Disposable {
 	public get isSqlCmdMode(): boolean {
 		return this._isSqlCmdMode;
 	}
+
+	public setState(oldState: QueryEditorState): void {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/common/editor/query/queryEditorInput.ts:110]: Shouldn't this be `newState`?

File: src/sql/workbench/common/editor/query/queryEditorInput.ts [hunk 11]
```diff
@@ -196,6 +204,11 @@ export abstract class QueryEditorInput extends EditorInput implements IConnectab
 		}
 	}
 
+	protected async notifyConnectionUriChanged(newUri: string): Promise<void> {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/common/editor/query/queryEditorInput.ts:207]: This is better named as `changeConnectionUri` since it's actually doing that action

File: src/sql/workbench/services/query/common/queryManagement.ts [hunk 12]
```diff
@@ -266,6 +268,15 @@ export class QueryManagementService implements IQueryManagementService {
 		});
 	}
 
+	public notifyConnectionUriChanged(newUri: string, oldUri: string): Promise<void> {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryManagement.ts:271]: This should also probably be `changeConnectionUri` since it's actually updating its own URI mappings - it's not just notifying that it was changed.

File: src/sql/workbench/services/query/common/queryModelService.ts [hunk 13]
```diff
@@ -47,6 +48,11 @@ export class QueryInfo {
 		this.queryEventQueue = [];
 		this.range = [];
 	}
+
+	public changeUri(newUri: string): void {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryModelService.ts:52]: Just have this be a setter 

```
public set uri(newUri: string) {
...
}
```

File: src/sql/workbench/services/query/common/queryModelService.ts [hunk 14]
```diff
@@ -409,6 +416,25 @@ export class QueryModelService implements IQueryModelService {
 		}
 	}
 
+	public async notifyConnectionUriChanged(newUri: string, oldUri: string): Promise<void> {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryModelService.ts:419]: This should also be `changeConnectionUri` - again since it's actually updating its own mapping in addition to notifying others to update themselves.

File: src/sql/workbench/services/query/common/queryModelService.ts [hunk 15]
```diff
@@ -409,6 +416,25 @@ export class QueryModelService implements IQueryModelService {
...
+		// Get existing query runner
+		let queryRunner = this.internalGetQueryRunner(oldUri);
+		if (!queryRunner) {
+			this._logService.error(`A Query and QueryRunner was not found for '${oldUri}'`);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryModelService.ts:423]: This should throw (or at least return to avoid null ref on 429)

File: src/sql/workbench/services/query/common/queryModelService.ts [hunk 16]
```diff
@@ -409,6 +416,25 @@ export class QueryModelService implements IQueryModelService {
...
+		}
+		else if (this._queryInfoMap.has(newUri)) {
+			this._logService.error(`New URI '${newUri}' already has query info associated with it.`);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryModelService.ts:426]: Same here, this should probably throw the error too

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/liveshare/src/providers/queryProvider.ts (hunk 3 above)
Comment (smartguest) [reply extensions/liveshare/src/providers/queryProvider.ts:140]: I'm following with what is being done with rebuildIntellisenseCache, which used the request format, I'll try to switch to notify instead and see if it breaks anything.

Comment (smartguest) [reply extensions/liveshare/src/providers/queryProvider.ts:140]: Also it needs to match the name of the azdata method call otherwise it will complain the function does not exist for queryprovider. I will need to change the function back to connectionUriChanged for the dataprotocol otherwise this won't work.

Comment (smartguest) [reply extensions/liveshare/src/providers/queryProvider.ts:140]: the specific error is: Object literal may only specify known properties and connectionurichanged does not exist in type QueryProvider. So either we call it connectionUriChanged or notifyConnectionUriChanged in both sqlops and here, but it cannot be different.

File: src/sql/workbench/common/editor/query/queryEditorInput.ts (hunk 10 above)
Comment (smartguest) [reply src/sql/workbench/common/editor/query/queryEditorInput.ts:110]: We're updating this with the state of the old editor, but since it is technically a new state, I'll change it to that.

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 17]
```diff
@@ -44,6 +48,26 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
...
+		return this.saveProcess(newEditorInput);
+
+	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		let newEditorInput = await this.text.saveAs(group, options);
+		return this.saveProcess(newEditorInput);
+	}
+
+	private async saveProcess(fileEditorInput: IEditorInput): Promise<IEditorInput | undefined> {
```
Comment (Charles-Gagnon) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:62]: A name like `createFileQueryEditorInput` seems like a more descriptive name here

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 18]
```diff
@@ -44,6 +48,26 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
...
 	}
 
+	override async save(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		let newEditorInput = await this.text.save(group, options);
+		return this.saveProcess(newEditorInput);
+
+	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		let newEditorInput = await this.text.saveAs(group, options);
```
Comment (Charles-Gagnon) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:58]: Please add comments here explaining why we need to manually create the file input. Something like 

```
Create our own FileQueryEditorInput wrapper here so that the existing state (connection, results, etc) can be transferred from this untitled input to the new file input.
```

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by smartguest** [comment]
I added support for file input as requested by @kburtram It turns out that I forgot to change the uri to provider map when doing the connection change (This is required to get the correct provider for the old uri when runAction is done for the changeconnectionuri otherwise it will try to run on the new input and fail.) The replacement needs to happen after the uri connection change process for queryModelService is complete as it still requires the oldUri provider map to work.

File: src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts [hunk 19]
```diff
@@ -86,4 +88,24 @@ export class FileQueryEditorInput extends QueryEditorInput {
...
 	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		// Create our own FileQueryEditorInput wrapper here so that the existing state (connection, results, etc) can be transferred from this input to the new input.
+		let newEditorInput = await this.text.saveAs(group, options);
+		if (newEditorInput.resource.toString(true) === this.uri) {
+			return newEditorInput;
+		}
+		else {
+			return this.createFileQueryEditorInput(newEditorInput);
```
[the remark this answers is not in the document]
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:99]: Unfortunately, during a saveAs, I get this specific error  that says trying to get a query/subset failed and that gridPanel cannot read property map of undefined. I'll see what I can do to change this, but it looks like for now we need to create a new editor input.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:99]: Can you remove your changes here for that and then do them in a follow up PR? There's lot here already and I'm worried we'll miss something if we keep trying to change and fix stuff in this current PR. And it seems like the current behavior at least wouldn't change with what you've done for saving an existing file - so as long as there's no regression I think a separate PR would be greatly preferred.

[the remark this answers is not in the document]
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:99]: This works after a recompilation, however the file path on the top doesn't match unless I switch back and forth, I'll find a way to change the header part.

File: src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts [hunk 20]
```diff
@@ -86,4 +88,24 @@ export class FileQueryEditorInput extends QueryEditorInput {
 	public override rename(group: GroupIdentifier, target: URI): IMoveResult {
```
[the remark this answers is not in the document]
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:88]: I have attempted to move the file via file explorer directly, but so far the rename is not triggering on the vscode side, I'll investigate this in a separate PR once this one is pushed in.

[the remark this answers is not in the document]
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:88]: I attempted to rename via filesystem, but it did not seem to trigger the rename call, nor is moving it to another location, rather the file is marked as deleted after a tab switch. The Query still works and gives results at least, so no errors so far.

File: src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts [hunk 21]
```diff
@@ -86,4 +88,22 @@ export class FileQueryEditorInput extends QueryEditorInput {
...
 		return this.text.rename(group, target);
 	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		// Handle saveAs when new location URI is different from the current URI.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:93]: Please be more specific - I'd suggest splitting this into two comments (one in the if and one in the else) that explain what each section is doing differently

File: src/sql/workbench/services/connection/browser/connectionManagementService.ts [hunk 22]
```diff
@@ -1069,6 +1069,17 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		return Promise.resolve(true);
 	}
 
+	/**
+	 * Replaces connection info uri with new uri.
+	 */
+	public changeConnectionUri(newUri: string, oldUri: string) {
+		this._connectionStatusManager.changeConnectionUri(newUri, oldUri);
+		if (this._uriToProvider[oldUri]) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionManagementService.ts:1077]: Any reason you're not treating this like the others where it being missing from the map would be considered an error?

**Comment by smartguest** [comment]
@Charles-Gagnon Added the correct dependencies and it seems to work fine, both issues seem to have been fixed with this PR. I'm not sure about the rename part since I can't get it to trigger, but there is no regression so far.

File: src/sql/workbench/services/query/common/queryManagement.ts [hunk 23]
```diff
@@ -266,6 +268,15 @@ export class QueryManagementService implements IQueryManagementService {
 		});
 	}
 
+	public changeConnectionUri(newUri: string, oldUri: string): Promise<void> {
+		let item = this._queryRunners.get(oldUri);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryManagement.ts:272]: Consider handling the uri not being there by throwing a more specific error

File: src/sql/workbench/services/query/common/queryManagement.ts [hunk 24]
```diff
@@ -266,6 +268,15 @@ export class QueryManagementService implements IQueryManagementService {
...
+		this._queryRunners.set(newUri, item);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryManagement.ts:273]: And if we have the new URI already that should also probably be an error like in queryModel

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
@alanrenmsft @kburtram Could you look at this when you get a chance? Would be good to make sure I didn't miss anything.

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 25]
```diff
@@ -44,6 +48,28 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
 		return this.text.model.hasAssociatedFilePath;
 	}
 
+	override async save(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		// Create our own FileQueryEditorInput wrapper here so that the existing state (connection, results, etc) can be transferred from this input to the new file input.
+		let newEditorInput = await this.text.save(group, options);
+		return this.createFileQueryEditorInput(newEditorInput);
+
```
Comment (alanrenmsft) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:55]: remove blank lines

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 26]
```diff
@@ -44,6 +48,28 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
 		return this.text.model.hasAssociatedFilePath;
 	}
 
+	override async save(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		// Create our own FileQueryEditorInput wrapper here so that the existing state (connection, results, etc) can be transferred from this input to the new file input.
```
Comment (alanrenmsft) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:52]: save and saveAs looks the same including the comments, you could have private method and both save and saveAs can call it.

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 27]
```diff
@@ -44,6 +48,28 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
 		return this.text.model.hasAssociatedFilePath;
 	}
 
+	override async save(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
```
Comment (alanrenmsft) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:51]: there are room for refactoring so that you don't have to repeat the same comment

File: src/sql/platform/connection/common/connectionStatusManager.ts [hunk 28]
```diff
@@ -49,6 +50,21 @@ export class ConnectionStatusManager {
 		return !!this.findConnection(id);
 	}
 
+	public changeConnectionUri(newId: string, oldId: string) {
```
Comment (alanrenmsft) [inline src/sql/platform/connection/common/connectionStatusManager.ts:53]: lets be consistent and rename the arguments to newUri and oldUri

Comment (smartguest) [reply src/sql/platform/connection/common/connectionStatusManager.ts:53]: I was being consistent with the rest of the file, but if we want it to be consistent with the other calls, sure thing.

**Comment by smartguest** [comment]
The already has provider check has been removed as it turns out that the provider is not deleted when a connection is closed, meaning that if you try to save on a file that already has a provider, the save as won't work, this has been removed for now.

File: src/sql/base/query/browser/untitledQueryEditorInput.ts [hunk 29]
```diff
@@ -44,6 +48,31 @@ export class UntitledQueryEditorInput extends QueryEditorInput implements IUntit
...
+	}
+
+	override async saveAs(group: GroupIdentifier, options?: ISaveOptions): Promise<IEditorInput | undefined> {
+		return this.createFileQueryEditorInput(group, options, false);
+	}
+
+	private async createFileQueryEditorInput(group: GroupIdentifier, options?: ISaveOptions, isRegularSave?: boolean): Promise<IEditorInput> {
+		// Create our own FileQueryEditorInput wrapper here so that the existing state (connection, results, etc) can be transferred from this input to the new file input.
+		let fileEditorInput: IEditorInput;
+		if (isRegularSave) {
```
Comment (Charles-Gagnon) [inline src/sql/base/query/browser/untitledQueryEditorInput.ts:62]: Why did you change this? What you had before was fine - moving it in here just means you have to add this extra logic instead of just passing in the editor input that each call makes itself.

Comment (smartguest) [reply src/sql/base/query/browser/untitledQueryEditorInput.ts:62]: I thought moving it into one would work, but I'll put it back.

File: src/sql/workbench/services/connection/browser/connectionManagementService.ts [hunk 30]
```diff
@@ -1069,6 +1069,20 @@ export class ConnectionManagementService extends Disposable implements IConnecti
...
+	/**
+	 * Replaces connection info uri with new uri.
+	 */
+	public changeConnectionUri(newUri: string, oldUri: string) {
+		this._connectionStatusManager.changeConnectionUri(newUri, oldUri);
+		if (!this._uriToProvider[oldUri]) {
+			this._logService.error(`No provider found for old URI : '${oldUri}'`);
+			throw new Error(nls.localize('connectionManagementService.noProviderForUri', 'Could not find provider for uri: {0}', oldUri));
+		}
+		// Provider will persist after disconnect, it is okay to overwrite the map if it exists from a previously deleted connection.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionManagementService.ts:1081]: Hm, not sure I like the behavior of the mapping persisting even after the editor is closed/disconnected. So something to keep an eye on if you continue to make changes here...we should ideally be cleaning up everything associated with a query once the editor is closed.

But it seems fine for this since that's existing behavior.

**Comment by smartguest** [comment]
The vscode side of things already saves the file during a save, but at least if you save over a file with an existing connection it no longer destroys the original file window with an existing connection and will now work with the updated text, but the new window won't display anything now. I added a message to let the user know of this.

**Comment by Charles-Gagnon** [comment]
Can you give repro steps for what the issue is? This doesn't sound like behavior we should have - and telling the user to "close and re-open their editor" is definitely something we should avoid if we can.

**Comment by smartguest** [comment]
@Charles-Gagnon The issue is that if you open a file and connect it, then open another query or file and try to save it over the first file, this will overwrite the file on the original window, but it still works. (this is by design from vscode, and we have way to retrieve the newUri  before it happens).

**Comment by smartguest** [comment]
If you remove the error handling though, it overwrites it without any issue seemingly and it can still run and stuff.

File: src/sql/platform/connection/common/connectionStatusManager.ts [hunk 31]
```diff
@@ -49,6 +50,21 @@ export class ConnectionStatusManager {
 		return !!this.findConnection(id);
 	}
 
+	public changeConnectionUri(newUri: string, oldUri: string) {
+		let info = this.findConnection(oldUri);
+		if (!info) {
+			this._logService.error(`No connection found associated with old URI : '${oldUri}'`);
+			throw new Error(nls.localize('connectionStatusManager.noConnectionForUri', 'Could not find connection with uri: {0}', oldUri));
+		}
+		if (this._connections[newUri]) {
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/connectionStatusManager.ts:59]: Is this also going to cause problems like the other one if we're saving over an existing file?

File: src/sql/workbench/services/query/common/queryManagement.ts [hunk 32]
```diff
@@ -266,6 +271,23 @@ export class QueryManagementService implements IQueryManagementService {
 		});
 	}
 
+	public changeConnectionUri(newUri: string, oldUri: string): Promise<void> {
+		let item = this._queryRunners.get(oldUri);
+		if (!item) {
+			this._logService.error(`No query runner found for old URI : '${oldUri}'`);
+			throw new Error(nls.localize('queryManagement.noQueryRunnerForUri', 'Could not find Query Runner for uri: {0}', oldUri));
+		}
+		if (this._queryRunners.get(newUri)) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryManagement.ts:280]: Same here about overwriting an existing connection

File: src/sql/workbench/services/query/common/queryModelService.ts [hunk 33]
```diff
@@ -409,6 +416,27 @@ export class QueryModelService implements IQueryModelService {
...
 	}
 
+	public async changeConnectionUri(newUri: string, oldUri: string): Promise<void> {
+		// Get existing query runner
+		let queryRunner = this.internalGetQueryRunner(oldUri);
+		if (!queryRunner) {
+			this._logService.error(`A Query and QueryRunner was not found for '${oldUri}'`);
+			throw new Error(nls.localize('queryModelService.noQueryFoundForUri', 'No Query found for {0}', oldUri));
+		}
+		else if (this._queryInfoMap.has(newUri)) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/query/common/queryModelService.ts:426]: And here with overwriting the new connection

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
Just want to make sure that the "existing URI" checks aren't going to cause problems when saving over an existing file that's opened in another editor.

File: src/sql/platform/connection/common/connectionStatusManager.ts (hunk 31 above)
Comment (alanrenmsft) [reply src/sql/platform/connection/common/connectionStatusManager.ts:59]: an existing file that is already opened in ADS

File: src/sql/platform/connection/common/connectionStatusManager.ts [hunk 34]
```diff
@@ -49,6 +50,21 @@ export class ConnectionStatusManager {
 		return !!this.findConnection(id);
 	}
 
+	public changeConnectionUri(newUri: string, oldUri: string) {
```
Comment (alanrenmsft) [inline src/sql/platform/connection/common/connectionStatusManager.ts:53]: return type

File: src/sql/platform/connection/common/connectionStatusManager.ts (hunk 31 above)
Comment (smartguest) [reply src/sql/platform/connection/common/connectionStatusManager.ts:59]: If its caught, the thing is designed to return the original input, and nothing is changed here.

File: src/sql/workbench/services/query/common/queryManagement.ts (hunk 32 above)
Comment (smartguest) [reply src/sql/workbench/services/query/common/queryManagement.ts:280]: It shouldn't reach this point as it would be caught in changeConnectionUri, but in the unlikely case it does, I'll add a catch statement to change the uri back to the old one.

File: src/sql/workbench/services/query/common/queryModelService.ts (hunk 33 above)
Comment (smartguest) [reply src/sql/workbench/services/query/common/queryModelService.ts:426]: So basically the execution will try and call QueryModelService's changeConnectionUri, if it fails there, nothing happens and the error handling in queryEditorInput will deal with the changes. If it fails in queryManagement.ts, nothing changes and it goes all the way back to queryEditorInput. At this point, it should be good to run the connectionUriChanged action on the query in STS.

File: src/sql/workbench/services/query/common/queryManagement.ts (hunk 32 above)
Comment (smartguest) [reply src/sql/workbench/services/query/common/queryManagement.ts:280]: See my comment for QueryModelService.ts on how an error in each step will work.

File: src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts [hunk 35]
```diff
@@ -86,4 +88,29 @@ export class FileQueryEditorInput extends QueryEditorInput {
...
+				await this.changeConnectionUri(newUri);
+				this._results.uri = newUri;
+				// Create a new FileQueryEditorInput with current results and state in order to trigger a rename for editor tab name.
+				// (Tab name won't refresh automatically if current input is reused directly)
+				let newFileQueryInput = this.instantiationService.createInstance(FileQueryEditorInput, '', (newEditorInput as FileEditorInput), this.results);
+				newFileQueryInput.state.setState(this.state);
+				return newFileQueryInput;
+			}
+			catch (error) {
+				return newEditorInput;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:112]: Comment explaining why we're returning the plain file input here and what that means for the editor.

But in general I don't like that we have something like this. Errors shouldn't be expected and so having fallback stuff like this seems to imply we think it would happen often enough we have to account for that. It'd be good to figure out why/when this case would be hit and try to fix those scenarios in the long term.

**Review by Charles-Gagnon — APPROVED** [review]
As mentioned in the comment, overall it makes me a little worried to see all these error checks everywhere - it seems like the chance of something not being reverted correctly and leaving things in a half-broken state is quite likely.

But fixing up that would be a much larger task so I guess this is ok for now, but let's make sure we spend a good amount of time testing various save scenarios once this gets in.

File: src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts (hunk 35 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/fileQueryEditorInput.ts:112]: I agree that it shouldn't be, and in fact it works without having the error checks altogether, but there needs to be a delay or some way to check that the file has updated properly before running the grid refresh on the file and this will require another PR and some time to investigate.

**Review by alanrenmsft — APPROVED** [review]
